### ISSUE TITLE
quantized mesh encoding?

### ISSUE THREAD

**Issue by justb4** [geo-data/cesium-terrain-builder#29]
Just wondering: it seems that the tiles have plain heights encoded (`TerrainTiler.cpp`?). If so, are there plans for quantized-mesh encoding as in http://cesiumjs.org/data-and-assets/terrain/formats/quantized-mesh-1.0.html? If not I/we may add this, though time is not on my side. We have very detailed (0.5m res) Dutch open height data: http://app.map5.nl/nltopo/#rd/relief_struct/12/196286.4/446444.9.

**Comment by justb4** [comment]
OK, I see it is in the TODO in the README now...

**Comment by homme** [comment]
Hi Just, Yes, it's in the TODO.  Unfortunately we don't have any funding or project requirements to implement this at the moment but welcome any external contributions.  I'm imagining implementing quantized mesh would require some refactoring to enable the different tile formats to be specified as plugins, as a sensible mesh implementation would differ considerably from the heightmap format.

I'll keep this issue open to track any progress.

**Comment by BarryReidPropeller** [comment]
Hi guys, would love to see this added as well!

**Comment by markerikson** [comment]
Searching through Github turns up https://github.com/geoadmin/3d-forge, which appears to have the ability to read and write quantized-mesh files.  Could be a useful source of info for someone trying to implement this feature in CTB.

**Comment by homme** [comment]
@markerikson Thanks Mark, that looks very interesting!

**Comment by BarryReidPropeller** [comment]
Anyone make any progress on this?

**Comment by homme** [comment]
@BarryReidPropeller I'm afraid not: this requires funding or a pull request...

**Comment by lapcchan** [comment]
@homme i am interested in funding the quantized-mesh implementation, you know of anyone interested in this project?

**Comment by chabok** [comment]
Hi Homme, we might be able to assist you in this regard, we have an application and would be happy to discuss it with you.

**Comment by markerikson** [comment]
One more useful-ish thing I just found.  This project looks like it uses a combination of GDAL and CGAL to turn a GDAL-supported raster into a Triangulated Irregular Network:  https://github.com/shashankkhare/refine/blob/master/TetrahedronBuilder.h.

Also, the 3DForge file I linked earlier has now been shuffled to https://github.com/geoadmin/3d-forge/blob/57b6ba5942bf0d9d9a6e3a6f4471a2e9f3a89c61/forge/terrain/__init__.py.  Note that they seem to be generating their TINs from shapefiles or something. 

I guess the other thought would be to do a DEM-to-shapefile conversion?  Something like http://gis.stackexchange.com/questions/28298/how-to-extract-raster-dem-elevations-to-vector-attribute-tables-in-arcgis-10 or http://freegeographytools.com/2007/converting-digital-elevation-models-to-shapefiledxf-contours.

It _seems_ like there's enough information out there to do this, if someone had the time to pull all the pieces together.  Cesium defines what the format should look like, the 3DForge project has some Python code that can read and write quantized-mesh files and translate TIN data into QM format, and there's techniques for turning raster data into a mesh or vector.  

It's a feature I'd love to see added to CTB, I just have no time or ability to try tackling it myself.  (Naturally.)

**Comment by loicgasser** [comment]
Hi everyone, I thinking about extracting this encoder/decoder from the project and publish it as a pypi module. I'll try to do it very soon and will let you know when it's ready.

**Comment by justb4** [comment]
@loicgasser +1 for pypi module. We may even look at integration with MapProxy http://mapproxy.org later.

**Comment by rspalko** [comment]
It might be useful to look at SCAPE also which is a C++ implementation of the paper by Michael Garland and Paul Heckbert at CMU.   (Sometimes technology comes full circle as we again need some of the same efficiency techniques for fast rendering on the web that we needed on the desktop.)  I was able to get scape to build on CentOS 6 and use it to convert a heightmap to TIN format.  This code could be ported and fed into the 3d-forge quantized terrain mesh writer I believe.  

Cesium has previously referenced Michael Garland's work in their presentations on Terrain services.
https://cesiumjs.org/presentations/Rendering%20the%20Whole%20Wide%20World%20on%20the%20World%20Wide%20Web.pdf

SCAPE download
https://www.cs.cmu.edu/~./garland/scape/

https://www.cs.cmu.edu/~./garland/scape/scape.pdf

**Comment by dg3feiko** [comment]
+1

**Comment by markerikson** [comment]
Someone just posted in the Cesium group that they've implemented a quantized mesh generator.  No specific details yet.  Link: https://groups.google.com/d/msg/cesium-dev/IpcBEvjt-DA/jtbIOATQEgAJ

**Comment by loicgasser** [comment]
pypi package for early birds is ready here: https://pypi.python.org/pypi/quantized-mesh-tile/
doc: https://quantized-mesh-tile.readthedocs.io
Hope this will be helpful

**Comment by markerikson** [comment]
Throwing out some more examples of relevant-looking links:

- https://github.com/geoadmin/3d-forge/issues/168
- http://blog.mastermaps.com/2016/09/creating-tin-from-raster-dem.html
- https://github.com/loicgasser/quantized-mesh-tile/issues/6
- https://github.com/rolker/cesium-tools
- https://github.com/bertt/awesome-quantized-mesh-tiles
- https://github.com/nmandery/conference-notes/blob/bae9d7d99cd648d96b49dc4b3b8c84421ff1d8f6/2016-06-23-geopython_2016/swisstopo_loic_gasser-quantized_mesh_encoder.md


Basically, looks like there's plenty of info on the format, and several examples of implementations for generating and writing QM.  Just need someone with the time, interest, and expertise to add it to CTB :)

**Comment by ahuarte47** [comment]
Hi, I have implemented the quantized-mesh output to the CTB (ctb-tile tool). It provides two new features:
- New quantized-mesh format ouput.
- New metadata file output.

It is free to test and use.

Details:
https://www.linkedin.com/pulse/fast-cesium-terrain-rendering-new-quantized-mesh-output-alvaro-huarte/

Pull request:
https://github.com/geo-data/cesium-terrain-builder/pull/64

**Comment by delfrrr** [comment]
Hi, we were working on a C++ [tool/lib](https://github.com/heremaps/tin-terrain) for conversion raster (GeoTIFF) heightmaps into [TIN](https://en.wikipedia.org/wiki/Triangulated_irregular_network)s (basically optimized mesh). Since we already generating a mesh we added a tiling and quantized-mesh encoding features!

The tool is MIT licensed and can be found at https://github.com/heremaps/tin-terrain

Also, there is [the example](https://github.com/heremaps/quantized-mesh-viewer) of rendering output with Cesium.js

I guess it addresses this issue and relates to #64 and #55 

Feedback appreciated!

**Comment by Hisos** [comment]
Hi, @ahuarte47 ctb-tile gets killed after "90..." (vs2015 Release x64, Debug is ok) in your code "ahuarte47/cesium-terrain-builder branch master-quantized-mesh". And if the output directory is not empty, both the debug and release will be killed at the beginning.
